### ISSUE TITLE
Cannot compile F# code that uses integer comparison

### ISSUE THREAD

**Issue by abrodersen** [dotnet/cli#692]
The following F# code fails to compile on @1f30ecc :
```
namespace Test
open System

module Program =
  [<EntryPoint>]
  let Main (args: string array) =
    if 1 < 2 then 0 else 1
```

Using this `project.json`:
```
{
  "version": "1.0.0-*",
  "dependencies": {
    "Microsoft.FSharp.Core.netcore": "1.0.0-alpha-*",
    "System.Runtime": "4.0.21-rc2-*"
  },
  "frameworks": {
    "dnxcore50": { }
  },
  "compilerName": "fsc",
  "compileFiles": [
    "bug.fs"
  ]
}
```

Giving this compiler error:
```
bug.fs(7,8): error FS0001: The type 'int' does not support the 'comparison' constraint. For example, it does not support the 'System.IComparable' interface

Compilation failed.
    0 Warning(s)
    1 Error(s)

Time elapsed 00:00:04.8821433
```

**Comment by enricosada** [comment]
i'll take a look, thx @abrodersen

**Comment by dsyme** [comment]
@enricosada What is the full ``fsc.exe`` command line being run? I suspect a bad assembly reference (e.g. an extraneous reference to mscorlib.dll) or a missing ``--targetprofile:netcore``.  

@abrodersen If you can find a way to get the full command line being generated by dotnet cli that would be helpful.

**Comment by dsyme** [comment]
From Dave Wecker:

> My test for this specific problem is VERY simple:

``` 
namespace TestApp

open System
open System.Diagnostics

module Program =

    [<EntryPoint>]
    let Main (args: string array) =
        if 1 < 2 then
            printfn "Got to FSHARP MAIN!!!"
        0
```

> With a JSON file that just has:

```
{
    "version": "1.0.0-*",
    "compilationOptions": {
        "emitEntryPoint": true
    },

    "dependencies": {
        "Microsoft.FSharp.Core.netcore": "1.0.0-*",
        "System.IO": "4.0.11-*",
        "System.Console": "4.0.0-*",
        "System.Runtime": "4.0.21-*",
        "System.Diagnostics.Process": "4.1.0-*",
        "Microsoft.NETCore.Runtime": "1.0.1-*"
    },

    "frameworks": {
        "dnxcore50": { }
    },
    
    "compilerName": "fsc",
    "compileFiles": [
        "Program.fs"
    ]
}

AFAICT, the fsc command line (using dotnet –v) looks like:

Running c:\depot\Git\wecker\dotnet-cli\artifacts\win7-x64\stage2\bin\corehost.exe "c:\depot\Git\wecker\dotnet-cli\artifacts\win7-x64\stage2\bin\fsc.exe" -d:DEBUG -d:TRACE -d:DNXCORE50 --target:exe --win32manifest:"c:\depot\Git\wecker\dotnet-cli\artifacts\win7-x64\stage2\bin\default.win32manifest" --noframework --nologo --simpleresolution --debug:full "C:\depot\GIT\wecker\dotnet\Test1\bin\.dotnetrun\4d6719f18fed49bbb651ad6f3656cbb5\dotnet-compile.assemblyinfo.fs" --out:"C:\depot\GIT\wecker\dotnet\Test1\bin\.dotnetrun\4d6719f18fed49bbb651ad6f3656cbb5\Test1.exe" -r:"C:\Users\wecker\.dnx\packages\Microsoft.FSharp.Core.netcore\1.0.0-alpha-151221\lib\DNXCore50\FSharp.Core.dll" -r:"C:\Users\wecker\.dnx\packages\Microsoft.Win32.Primitives\4.0.0\ref\dotnet\Microsoft.Win32.Primitives.dll" -r:"C:\Users\wecker\.dnx\packages\System.Collections\4.0.11-rc2-23616\ref\dotnet5.4\System.Collections.dll" -r:"C:\Users\wecker\.dnx\packages\System.Collections.Concurrent\4.0.11-rc2-23616\ref\dotnet5.4\System.Collections.Concurrent.dll" -r:"C:\Users\wecker\.dnx\packages\System.Collections.NonGeneric\4.0.0\ref\dotnet\System.Collections.NonGeneric.dll" -r:"C:\Users\wecker\.dnx\packages\System.Collections.Specialized\4.0.0\ref\dotnet\System.Collections.Specialized.dll" -r:"C:\Users\wecker\.dnx\packages\System.ComponentModel.EventBasedAsync\4.0.10\ref\dotnet\System.ComponentModel.EventBasedAsync.dll" -r:"C:\Users\wecker\.dnx\packages\System.Console\4.0.0-rc2-23707\ref\dotnet5.4\System.Console.dll" -r:"C:\Users\wecker\.dnx\packages\System.Diagnostics.Debug\4.0.11-rc2-23616\ref\dotnet5.4\System.Diagnostics.Debug.dll" -r:"C:\Users\wecker\.dnx\packages\System.Diagnostics.Process\4.1.0-rc2-23707\ref\dotnet5.5\System.Diagnostics.Process.dll" -r:"C:\Users\wecker\.dnx\packages\System.Diagnostics.Tools\4.0.1-rc2-23616\ref\dotnet5.1\System.Diagnostics.Tools.dll" -r:"C:\Users\wecker\.dnx\packages\System.Diagnostics.Tracing\4.0.20\ref\dotnet\System.Diagnostics.Tracing.dll" -r:"C:\Users\wecker\.dnx\packages\System.Globalization\4.0.11-rc2-23616\ref\dotnet5.4\System.Globalization.dll" -r:"C:\Users\wecker\.dnx\packages\System.Globalization.Extensions\4.0.0\ref\dotnet\System.Globalization.Extensions.dll" -r:"C:\Users\wecker\.dnx\packages\System.IO\4.0.11-rc2-23707\ref\dotnet5.4\System.IO.dll" -r:"C:\Users\wecker\.dnx\packages\System.IO.FileSystem\4.0.0\ref\dotnet\System.IO.FileSystem.dll" -r:"C:\Users\wecker\.dnx\packages\System.IO.FileSystem.Primitives\4.0.0\ref\dotnet\System.IO.FileSystem.Primitives.dll" -r:"C:\Users\wecker\.dnx\packages\System.Linq\4.0.1-rc2-23616\ref\dotnet5.1\System.Linq.dll" -r:"C:\Users\wecker\.dnx\packages\System.Linq.Expressions\4.0.11-rc2-23616\ref\dotnet5.4\System.Linq.Expressions.dll" -r:"C:\Users\wecker\.dnx\packages\System.Linq.Queryable\4.0.1-rc2-23616\ref\dotnet5.1\System.Linq.Queryable.dll" -r:"C:\Users\wecker\.dnx\packages\System.Net.Primitives\4.0.10\ref\dotnet\System.Net.Primitives.dll" -r:"C:\Users\wecker\.dnx\packages\System.Net.Requests\4.0.11-rc2-23616\ref\dotnet5.4\System.Net.Requests.dll" -r:"C:\Users\wecker\.dnx\packages\System.Net.WebHeaderCollection\4.0.0\ref\dotnet\System.Net.WebHeaderCollection.dll" -r:"C:\Users\wecker\.dnx\packages\System.Reflection\4.1.0-rc2-23616\ref\dotnet5.4\System.Reflection.dll" -r:"C:\Users\wecker\.dnx\packages\System.Reflection.Emit\4.0.1-rc2-23616\ref\dotnet5.2\System.Reflection.Emit.dll" -r:"C:\Users\wecker\.dnx\packages\System.Reflection.Emit.ILGeneration\4.0.1-rc2-23616\ref\dotnet5.1\System.Reflection.Emit.ILGeneration.dll" -r:"C:\Users\wecker\.dnx\packages\System.Reflection.Extensions\4.0.1-rc2-23616\ref\dotnet5.1\System.Reflection.Extensions.dll" -r:"C:\Users\wecker\.dnx\packages\System.Reflection.Primitives\4.0.0\ref\dotnet\System.Reflection.Primitives.dll" -r:"C:\Users\wecker\.dnx\packages\System.Reflection.TypeExtensions\4.1.0-rc2-23616\ref\dotnet5.4\System.Reflection.TypeExtensions.dll" -r:"C:\Users\wecker\.dnx\packages\System.Resources.ResourceManager\4.0.1-rc2-23616\ref\dotnet5.1\System.Resources.ResourceManager.dll" -r:"C:\Users\wecker\.dnx\packages\System.Runtime\4.0.21-rc2-23707\ref\dotnet5.4\System.Runtime.dll" -r:"C:\Users\wecker\.dnx\packages\System.Runtime.Extensions\4.0.11-rc2-23616\ref\dotnet5.4\System.Runtime.Extensions.dll" -r:"C:\Users\wecker\.dnx\packages\System.Runtime.Handles\4.0.0\ref\dotnet\System.Runtime.Handles.dll" -r:"C:\Users\wecker\.dnx\packages\System.Runtime.InteropServices\4.0.21-rc2-23616\ref\dotnet5.4\System.Runtime.InteropServices.dll" -r:"C:\Users\wecker\.dnx\packages\System.Runtime.Loader\4.0.0-rc2-23616\ref\dotnet5.5\System.Runtime.Loader.dll" -r:"C:\Users\wecker\.dnx\packages\System.Runtime.Numerics\4.0.1-rc2-23616\ref\dotnet5.2\System.Runtime.Numerics.dll" -r:"C:\Users\wecker\.dnx\packages\System.Text.Encoding\4.0.11-rc2-23616\ref\dotnet5.4\System.Text.Encoding.dll" -r:"C:\Users\wecker\.dnx\packages\System.Text.Encoding.Extensions\4.0.11-rc2-23616\ref\dotnet5.4\System.Text.Encoding.Extensions.dll" -r:"C:\Users\wecker\.dnx\packages\System.Text.RegularExpressions\4.0.11-rc2-23616\ref\dotnet5.4\System.Text.RegularExpressions.dll" -r:"C:\Users\wecker\.dnx\packages\System.Threading\4.0.11-rc2-23616\ref\dotnet5.4\System.Threading.dll" -r:"C:\Users\wecker\.dnx\packages\System.Threading.Overlapped\4.0.0\ref\dotnet\System.Threading.Overlapped.dll" -r:"C:\Users\wecker\.dnx\packages\System.Threading.Tasks\4.0.11-rc2-23616\ref\dotnet5.4\System.Threading.Tasks.dll" -r:"C:\Users\wecker\.dnx\packages\System.Threading.Tasks.Parallel\4.0.1-rc2-23616\ref\dotnet5.2\System.Threading.Tasks.Parallel.dll" -r:"C:\Users\wecker\.dnx\packages\System.Threading.Thread\4.0.0-rc2-23616\ref\dotnet5.4\System.Threading.Thread.dll" -r:"C:\Users\wecker\.dnx\packages\System.Threading.ThreadPool\4.0.10-rc2-23616\ref\dotnet5.4\System.Threading.ThreadPool.dll" -r:"C:\Users\wecker\.dnx\packages\System.Threading.Timer\4.0.1-rc2-23616\ref\dotnet5.3\System.Threading.Timer.dll" "C:/depot/GIT/wecker/dotnet/Test1/Program.fs"

```

**Comment by dsyme** [comment]
@enricosada Do you need a ``--targetprofile:netcore`` flag?

**Comment by davidfowl** [comment]
@dsyme What does that do?

**Comment by enricosada** [comment]
yes @dsyme , it's missing

**Comment by enricosada** [comment]
to debug arguments, you can add a `-t <directory>` to use a temp directory.
in the temp directory there is the response file used by fsc with arguments

**Comment by davidfowl** [comment]
Just look in the obj dir, it's there by default after a build.

**Comment by enricosada** [comment]
better :smile: , thx @davidfowl

**Comment by KevinRansom** [comment]
@davidfowl @enricosada @dsyme The fsharp compiler command line currently needs --targetprofile:netcore --noframework in order to target a portable profile or the coreclr.

This is a legacy of the desktop compiler.

**Comment by dsyme** [comment]
@davidfowl Roughly speaking it tells the compiler to look in System.Runtime instead of mscorlib for basic types.

**Comment by davidfowl** [comment]
@dsyme why not instead mimic what C# does and allow passing `/noconfig` and just handing over all of the references without an assumption about `mscorlib` or `System.Runtime`? I guess there must be an assumption somewhere about which assembly contains the intrinsic definitions...

**Comment by KevinRansom** [comment]
@davidfowl I'm sure we will end up doing something with the command line to make it less ungainly, the suggested model seems perfectly workable.  The legacy that is the current command line options have been in the product for a long time, and will need to continue to work.  We currently have a bunch of higher priority work than cleaning up the cl options.  Especially since the dotnet cl tool will so nicely hide that for now.
As an FYI, things that have a slightly higher priority:
   Signing assemblies, 
   Symbol generation,
   A new abstraction for type provider referencing types known to the compiler.
 
I hope this helps

**Comment by enricosada** [comment]
i'll send a fix monday or tuesday or update this issue, lots of build problems this week

**Comment by dsyme** [comment]
@enricosada Note that if "dotnet cli" adds this switch by default, then it won't work for .NET Framework compilation.   I suppose you need to make sure you propagate an "other flags" setting, which would be useful in any case.

**Comment by davidfowl** [comment]
You can check the target framework before adding the switch

**Comment by enricosada** [comment]
I'll do like that, thx @davidfowl

**Comment by dsyme** [comment]
Thanks @davidfowl!

**Comment by dbwz8** [comment]
Another question on this. If I build and run a test without int comparisons (and without the above switches), everything works. If I build and run *with* int comparisons *and* the above switches, the build is fine, but the result of running the .exe is that it can't find any libraries (see below). What's the right way to make this happen correctly? Note, if I do a **dotnet publish** (which copies all the libraries locally), it does work. It's just not the right solution.
```
> bin\Debug\dnxcore50\Test1.exe

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'FSharp.Core, Version=4.4.0.9055, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
   at System.ModuleHandle.ResolveType(RuntimeModule module, Int32 typeToken, IntPtr* typeInstArgs, Int32 typeInstCount, IntPtr* methodInstArgs, Int32 methodInstCount, ObjectHandleOnStack type)
   at System.ModuleHandle.ResolveTypeHandleInternal(RuntimeModule module, Int32 typeToken, RuntimeTypeHandle[] typeInstantiationContext, RuntimeTypeHandle[] methodInstantiationContext)
   at System.Reflection.RuntimeModule.ResolveType(Int32 metadataToken, Type[] genericTypeArguments, Type[] genericMethodArguments)
   at System.Reflection.CustomAttribute.FilterCustomAttributeRecord(CustomAttributeRecord caRecord, MetadataImport scope, Assembly& lastAptcaOkAssembly, RuntimeModule decoratedModule, MetadataToken decoratedToken, RuntimeType attributeFilterType, Boolean mustBeInheritable, Object[] attributes, IList derivedAttributes, RuntimeType& attributeType, IRuntimeMethodInfo& ctor, Boolean& ctorHasParameters, Boolean& isVarArg)
   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeModule decoratedModule, Int32 decoratedMetadataToken, Int32 pcaCount, RuntimeType attributeFilterType, Boolean mustBeInheritable, IList derivedAttributes, Boolean isDecoratedTargetSecurityTransparent)
   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeAssembly assembly, RuntimeType caType)
   at System.Attribute.GetCustomAttributes(Assembly element, Type attributeType, Boolean inherit)
   at System.AppDomain.GetTargetFrameworkName()
```

**Comment by KevinRansom** [comment]
Is it not the right solution?   I thought that test and publish were how you deployed runnable apps, am I wrong?

**Comment by enricosada** [comment]
@dbwz8 you need FSharp.Core in the same path as the compiled .exe, otherwise then you run the exe the dotnet cannot find the fsharp.core assembly.

dotnet build **compile and add reference inside output directory**, so should work after #894 it's merged.
Manually you should compile and copy the required referenced assemblies.

The dotnet publish copy the whole coreclr, with the compiled exe and references, readi to xcopy

**Comment by enricosada** [comment]
that's fixed with #1285 (already in the binaries) and can be closed /cc @blackdwarf
